pkg/vmprovider/providers/vsphere/vmprovider_vm_utils.go

```diff
@@ -440,3 +438,28 @@ func resolveContentLibraryUUID(
 
 	return libraryUUID, nil
 }
+
+// HardwareVersionForPVCandPCIDevices returns a hardware version for VMs with PVCs and PCI devices(vGPUs/DDPIO devices)
+// The hardware version is determined based on the below criteria: VMs with
+// - Persistent Volume Claim (PVC) get the max(the image hardware version, minimum supported virtual hardware version for persistent volumes)
+// - vGPUs/DDPIO devices get the max(the image hardware version, minimum supported virtual hardware version for PCI devices)
+// - Both vGPU/DDPIO devices and PVCs get the max(the image hardware version, minimum supported virtual hardware version for PCI devices)
+// - none of the above returns 0.
+func HardwareVersionForPVCandPCIDevices(vmSpec vmopv1.VirtualMachineSpec, imageHWVersion int32, configSpec *types.VirtualMachineConfigSpec) int32 {
```

Comment:
This MR handles `only` VMs that have devices like PVCs/PCI devices when a class config spec hardware version is unset.
All other VMs when a class config spec hardware version isn't set, will still default to the cluster default which we will need to document as a behavior change after this FSS.

So, my question was this floorHardwareVersion logic will not apply to VMs that don't have PVCs and PCI devices (vGPU && DDPIO bespoke) because we decided not to default everything to the image hardware version with this MR.

Planning to take this is up as a follow up if required?. does that sound good. I'll make all the other requested changes for now

Reply:
nit: `s/This MR/This PR/g`

Your GitLab is showing :smile:

> Planning to take this is up as a follow up if required?. does that sound good. I'll make all the other requested changes for now

Sure, please file a GH issue with an explanation of the change you will handle in the future.

Reply:
Will do. Need some more context before filing an issue. 

> if the VM Op VM does not have PVCs then floorHardwareVersion is imageHWVersion
Do we plan on flooring all VMs with `no PVC` and `no vGPUs/DDPIO` with `no class config spec hardware version` set to the image's hardwareVersion?.  I thought we decided against it and said they will default to the cluster's version because the Image's version doesn't necessarily describe the class's devices anymore.

Reply:
> Do we plan on flooring all VMs with no PVC and no vGPUs/DDPIO with no class config spec hardware version set to the image's hardwareVersion?. I thought we decided against it and said they will default to the cluster's version because the Image's version doesn't necessarily describe the class's devices anymore.

No clue. That's why I suggest the `GetMinimumCompatibleHardwareVersion` function take floor as an input param. The logic for what consitutes the floor value we send in is determined in our business logic. The logic in the `GetMinimumCompatibleHardwareVersion` function itself is based purely on what vSphere requires.

Reply:
FWIW, the `GetMinimumCompatibleHardwareVersion` function as I imagined it would be:

```go
func GetMinimumCompatibleHardwareVersion(
  ctx context.Context,
  floorHardwareVersion int32,
  configSpec types.VirtualMachineConfigSpec) int32
```

The purpose of the `floorHardwareVersion` as an input parameter is so the function can basically return the authoritative hardware version for the provided `configSpec` without needing logic after this function is called. However, I'm also fine if we wanted to do this:

```go
func GetMinimumCompatibleHardwareVersion(
  ctx context.Context,
  configSpec types.VirtualMachineConfigSpec) int32
```

The above _only_ looks at the provided `configSpec` to determine the hardware version and expects all other logic to occur before or after the function is called. Honestly, this probably makes more sense and keeps things isolated in a neater manner.